packages/react-components/react-virtualizer/src/hooks/useVirtualizerMeasure.ts

```diff
@@ -0,0 +1,102 @@
+import * as React from 'react';
+import { canUseDOM } from '@fluentui/react-utilities';
+import { useFluent_unstable as useFluent } from '@fluentui/react-shared-contexts';
+
+/**
+ * React hook that measures virtualized space based on a static size to ensure optimized virtualization length.
+ */
+export const useStaticVirtualizerMeasure = (
+  defaultItemSize: number,
+  scrollView: HTMLElement | null = null,
+  direction: 'vertical' | 'horizontal' = 'vertical',
+): {
+  virtualizerLength: number;
+  bufferItems: number;
+  bufferSize: number;
+} => {
+  const [virtualizerLength, setVirtualizerLength] = React.useState(0);
+  const [virtualizerBufferItems, setVirtualizerBufferItems] = React.useState(0);
+  const [virtualizerBufferSize, setVirtualizerBufferSize] = React.useState(0);
+
+  const { targetDocument } = useFluent();
+
+  // If no ref is provided, the virtualizer should be in the main body document's scroll.
+  const _scrollView = scrollView ?? targetDocument?.documentElement ?? null;
+
+  const container = React.useRef<HTMLElement | null>(null);
+
+  // the handler for resize observer
+  const handleResize = React.useCallback(() => {
+    const containerSize =
+      direction === 'vertical'
+        ? container.current?.getBoundingClientRect().height
+        : container.current?.getBoundingClientRect().width;
+
+    if (!containerSize) {
+      // Error? ignore?
+      return;
+    }
+
+    /*
+     * Number of items required to cover viewport.
+     */
+    const length = Math.ceil(containerSize / defaultItemSize + 1);
+
+    /*
+     * Number of items to append at each end, i.e. 'preload' each side before entering view.
+     */
+    const bufferItems = Math.max(Math.floor(length / 4), 2);
+
+    /*
+     * This is how far we deviate into the bufferItems to detect a redraw.
+     */
+    const bufferSize = Math.max(Math.floor((length / 8) * defaultItemSize), 1);
+
+    const totalLength = length + bufferItems * 2 + 1;
+
+    setVirtualizerLength(totalLength);
+    setVirtualizerBufferSize(bufferSize);
+    setVirtualizerBufferItems(bufferItems);
+  }, [defaultItemSize, direction]);
+
+  // Keep the reference of ResizeObserver in the state, as it should live through renders
+  const [resizeObserver] = React.useState(canUseDOM() ? new ResizeObserver(handleResize) : undefined);
```

Comment:
resize observers can trigger a lot and very often, consider debouncing the updates since react state updates + effects are synchronous

https://github.com/microsoft/fluentui/blob/master/packages/react-components/priority-overflow/src/debounce.ts

Reply:
additionally calling `getBoundingClientRect` will cause reflow which makes even more sense to debounce this until the javascript thread is idle

Reply:
Sounds good - I've put this as a local function in utilities as I wasn't sure we want the increased dependency just for this especially if we move this to an external package.